### PR TITLE
Asp6sample

### PR DESCRIPTION
**Pull request by xavierjohn**
ASP .NET 6 Sample.

### CODE REVIEW

File: examples/AspNetCore/Examples.AspNetCore.csproj [hunk 1]
```diff
@@ -1,28 +1,24 @@
...
     <ProjectReference Include="$(RepoRoot)\src\OpenTelemetry.Exporter.Console\OpenTelemetry.Exporter.Console.csproj" />
-    <ProjectReference Include="$(RepoRoot)\src\OpenTelemetry.Exporter.Jaeger\OpenTelemetry.Exporter.Jaeger.csproj" />
-    <ProjectReference Include="$(RepoRoot)\src\OpenTelemetry.Exporter.Zipkin\OpenTelemetry.Exporter.Zipkin.csproj" />
     <ProjectReference Include="$(RepoRoot)\src\OpenTelemetry.Exporter.OpenTelemetryProtocol\OpenTelemetry.Exporter.OpenTelemetryProtocol.csproj" />
     <ProjectReference Include="$(RepoRoot)\src\OpenTelemetry.Exporter.OpenTelemetryProtocol.Logs\OpenTelemetry.Exporter.OpenTelemetryProtocol.Logs.csproj" />
-    <ProjectReference Include="$(RepoRoot)\src\OpenTelemetry.Exporter.Prometheus\OpenTelemetry.Exporter.Prometheus.csproj" />
     <ProjectReference Include="$(RepoRoot)\src\OpenTelemetry.Instrumentation.AspNetCore\OpenTelemetry.Instrumentation.AspNetCore.csproj" />
     <ProjectReference Include="$(RepoRoot)\src\OpenTelemetry.Instrumentation.Http\OpenTelemetry.Instrumentation.Http.csproj" />
+    <ProjectReference Include="..\..\src\OpenTelemetry.Exporter.Jaeger\OpenTelemetry.Exporter.Jaeger.csproj" />
```
Comment (cijothomas) [inline examples/AspNetCore/Examples.AspNetCore.csproj:20]:
```suggestion
    <ProjectReference Include="$(RepoRoot)\src\OpenTelemetry.Exporter.Jaeger\OpenTelemetry.Exporter.Jaeger.csproj" />
```

File: examples/AspNetCore/Program.cs [hunk 2]
```diff
@@ -14,67 +14,131 @@
...
-    public class Program
+    "jaeger" => ResourceBuilder.CreateDefault().AddService(builder.Configuration.GetValue<string>("Jaeger:ServiceName"), serviceVersion: assemblyVersion, serviceInstanceId: Environment.MachineName),
+    "zipkin" => ResourceBuilder.CreateDefault().AddService(builder.Configuration.GetValue<string>("Zipkin:ServiceName"), serviceVersion: assemblyVersion, serviceInstanceId: Environment.MachineName),
+    "otlp" => ResourceBuilder.CreateDefault().AddService(builder.Configuration.GetValue<string>("Otlp:ServiceName"), serviceVersion: assemblyVersion, serviceInstanceId: Environment.MachineName),
+    _ => ResourceBuilder.CreateDefault().AddService(serviceName, serviceVersion: assemblyVersion, serviceInstanceId: Environment.MachineName),
+};
+
+builder.Logging.ClearProviders();
+
+builder.Logging.AddOpenTelemetry(options =>
```
Comment (cijothomas) [inline examples/AspNetCore/Program.cs:45]: https://github.com/open-telemetry/opentelemetry-dotnet/pull/2989/files#diff-42b554e34c7e926f354d61c80d72526bad109341f7ce5167ec6ada03f01a529fL72
Missing configuring `OpenTelemetryLoggerOptions` from existing example

File: examples/AspNetCore/Program.cs [hunk 3]
```diff
@@ -14,67 +14,131 @@
...
+        case "otlp":
+            options.AddOtlpExporter();
+            break;
+        default:
+            options.AddConsoleExporter();
+            break;
+    }
+});
+
+builder.Services.Configure<AspNetCoreInstrumentationOptions>(builder.Configuration.GetSection("AspNetCoreInstrumentation"));
```
Comment (cijothomas) [inline examples/AspNetCore/Program.cs:119]: Missing options from existing example:
```// For options which can be configured from code only.
                    services.Configure<AspNetCoreInstrumentationOptions>(options =>
```

**Review by cijothomas — CHANGES_REQUESTED** [review]
Please retain all the options from existing example. If removing anything intentionally done, please add a note with reasoning.

File: examples/AspNetCore/README.md [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+`ResourceBuilder` is associated with OpenTelemetry to associate the
+service name, version and the machine on which this program is running.
+
+The sample rate is set to emit all the traces using `AlwaysOnSampler`.
+You can try out different samplers like `TraceIdRatioBasedSampler`.
+
+## References
+
+* [ASP.NET Core 3.1 Example](https://github.com/open-telemetry/opentelemetry-dotnet/tree/core-1.1.0/examples/AspNetCore)
```
Comment (cijothomas) [inline examples/AspNetCore/README.md:18]: must use a commit just before this PR. 1.1.0 is too old

File: examples/AspNetCore/Program.cs (hunk 3 above)
Comment (xavierjohn) [reply examples/AspNetCore/Program.cs:119]: This code fails saying the statement is aways true (req.Request.Host != null;).
 // For options which can be configured from code only.
                    services.Configure<AspNetCoreInstrumentationOptions>(options =>
                    {
                        options.Filter = (req) =>
                        {
                            return req.Request.Host != null;
                        };
                    });

File: examples/AspNetCore/Program.cs (hunk 2 above)
Comment (xavierjohn) [reply examples/AspNetCore/Program.cs:45]: OpenTelemetryLoggerOptions  added

File: examples/AspNetCore/Examples.AspNetCore.csproj [hunk 5]
```diff
@@ -1,28 +1,24 @@
...
-    <IsPackable>false</IsPackable>
+    <TargetFramework>net6.0</TargetFramework>
+    <Nullable>enable</Nullable>
+    <ImplicitUsings>enable</ImplicitUsings>
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.Extensions.Http" Version="3.1.20" />
-    <PackageReference Include="Swashbuckle.AspNetCore" Version="$(SwashbuckleAspNetCorePkgVer)" />
+    <PackageReference Include="Swashbuckle.AspNetCore" Version="6.2.3" />
```
Comment (cijothomas) [inline examples/AspNetCore/Examples.AspNetCore.csproj:10]: Please update and use $(SwashbuckleAspNetCorePkgVer)

File: examples/AspNetCore/Program.cs [hunk 6]
```diff
@@ -14,67 +14,147 @@
...
-
-        public static IHostBuilder CreateHostBuilder(string[] args) =>
-            Host.CreateDefaultBuilder(args)
-                .ConfigureWebHostDefaults(webBuilder =>
-                {
-                    webBuilder.UseStartup<Startup>();
-                })
-                .ConfigureLogging((context, builder) =>
+        case "otlp":
+            options.AddOtlpExporter();
```
Comment (cijothomas) [inline examples/AspNetCore/Program.cs:52]: this breaks existing example which uses endpoint from config.

File: examples/AspNetCore/Program.cs [hunk 7]
```diff
@@ -14,67 +14,147 @@
...
+    app.UseSwaggerUI();
 }
+
+app.UseHttpsRedirection();
+
+app.UseAuthorization();
+
+app.MapControllers();
+
+app.Run();
```
Comment (cijothomas) [inline examples/AspNetCore/Program.cs:160]: missing prometheus route mapping.

File: examples/AspNetCore/Examples.AspNetCore.csproj [hunk 8]
```diff
@@ -1,28 +1,24 @@
 <Project Sdk="Microsoft.NET.Sdk.Web">
 
   <PropertyGroup>
-    <TargetFramework>netcoreapp3.1</TargetFramework>
-    <GenerateDocumentationFile>true</GenerateDocumentationFile>
-    <NoWarn>$(NoWarn);1591</NoWarn>
-    <IsPackable>false</IsPackable>
+    <TargetFramework>net6.0</TargetFramework>
+    <Nullable>enable</Nullable>
+    <ImplicitUsings>enable</ImplicitUsings>
```
[the remark this answers is not in the document]
Comment (xavierjohn) [reply examples/AspNetCore/Examples.AspNetCore.csproj:6]: That is the default when creating a new project. IsPackable is for creating NuGet packages and not needed here.

File: examples/AspNetCore/Program.cs (hunk 6 above)
Comment (xavierjohn) [reply examples/AspNetCore/Program.cs:52]: EndPoint is still coming from config.
`
        case "otlp":
            options.AddOtlpExporter(otlpOptions =>
                {
                    otlpOptions.Endpoint = new Uri(builder.Configuration.GetValue<string>("Otlp:Endpoint"));
                });
            break;
`

File: examples/AspNetCore/Program.cs [hunk 9]
```diff
@@ -14,67 +14,147 @@
...
+            break;
+
+        default:
+            options.AddConsoleExporter();
+
+            break;
     }
+});
+
+builder.Services.Configure<OpenTelemetryLoggerOptions>(opt =>
```
[the remark this answers is not in the document]
Comment (xavierjohn) [reply examples/AspNetCore/Program.cs:99]: Gives warning.
`warning SA1124: Do not use regions`

File: examples/AspNetCore/Program.cs (hunk 7 above)
Comment (xavierjohn) [reply examples/AspNetCore/Program.cs:170]: Good catch. Added.

File: examples/AspNetCore/Program.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (xavierjohn) [reply examples/AspNetCore/Program.cs:104]: I have ordered it as trace, logs and metrics

File: examples/AspNetCore/Examples.AspNetCore.csproj (hunk 5 above)
Comment (xavierjohn) [reply examples/AspNetCore/Examples.AspNetCore.csproj]: done

File: examples/AspNetCore/Examples.AspNetCore.csproj [hunk 10]
```diff
@@ -1,28 +1,24 @@
...
-    <ProjectReference Include="$(RepoRoot)\src\OpenTelemetry.Exporter.Zipkin\OpenTelemetry.Exporter.Zipkin.csproj" />
     <ProjectReference Include="$(RepoRoot)\src\OpenTelemetry.Exporter.OpenTelemetryProtocol\OpenTelemetry.Exporter.OpenTelemetryProtocol.csproj" />
     <ProjectReference Include="$(RepoRoot)\src\OpenTelemetry.Exporter.OpenTelemetryProtocol.Logs\OpenTelemetry.Exporter.OpenTelemetryProtocol.Logs.csproj" />
-    <ProjectReference Include="$(RepoRoot)\src\OpenTelemetry.Exporter.Prometheus\OpenTelemetry.Exporter.Prometheus.csproj" />
     <ProjectReference Include="$(RepoRoot)\src\OpenTelemetry.Instrumentation.AspNetCore\OpenTelemetry.Instrumentation.AspNetCore.csproj" />
     <ProjectReference Include="$(RepoRoot)\src\OpenTelemetry.Instrumentation.Http\OpenTelemetry.Instrumentation.Http.csproj" />
+    <ProjectReference Include="$(RepoRoot)\src\OpenTelemetry.Exporter.Jaeger\OpenTelemetry.Exporter.Jaeger.csproj" />
+    <ProjectReference Include="..\..\src\OpenTelemetry.Exporter.Prometheus\OpenTelemetry.Exporter.Prometheus.csproj" />
+    <ProjectReference Include="..\..\src\OpenTelemetry.Exporter.Zipkin\OpenTelemetry.Exporter.Zipkin.csproj" />
```
Comment (cijothomas) [inline examples/AspNetCore/Examples.AspNetCore.csproj:22]: PLease use ${RepoRoot}

File: examples/AspNetCore/Program.cs [hunk 11]
```diff
@@ -14,67 +14,157 @@
...
+builder.Logging.ClearProviders();
+
+builder.Logging.AddOpenTelemetry(options =>
+{
+    options.SetResourceBuilder(resourceBuilder);
+    var logExporter = builder.Configuration.GetValue<string>("UseLogExporter").ToLowerInvariant();
+    switch (logExporter)
+    {
+        case "otlp":
+            options.AddOtlpExporter();
```
Comment (cijothomas) [inline examples/AspNetCore/Program.cs:96]: this is missing endpoint reading from config file? its part of current example.

File: examples/AspNetCore/Examples.AspNetCore.csproj (hunk 10 above)
Comment (xavierjohn) [reply examples/AspNetCore/Examples.AspNetCore.csproj]: Fixed

File: examples/AspNetCore/Program.cs (hunk 11 above)
Comment (xavierjohn) [reply examples/AspNetCore/Program.cs:96]: End point reading is still there.

**Review by cijothomas — APPROVED** [review]

**Pull request merged**